src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -724,7 +724,20 @@ void BaseCouplingScheme::advanceTXTWriters()
 
 bool BaseCouplingScheme::reachedEndOfTimeWindow()
 {
-  return math::equals(getNextTimeStepMaxSize(), 0.0, _eps) || not hasTimeWindowSize();
+  if (math::equals(getNextTimeStepMaxSize(), 0.0, _minTimeStepSize) && getNextTimeStepMaxSize() > math::NUMERICAL_ZERO_DIFFERENCE && hasTimeWindowSize()) {
+    for (auto &data : _allData | boost::adaptors::map_values) {
+      auto   timeStepStorage = data->timeStepsStorage();
+      auto   values          = timeStepStorage.getTimesAndValues();
+      auto   ts              = values.first;
+      double endOfTimeWindow = getTime() - _computedTimeWindowPart + _timeWindowSize;
+      if (ts(ts.size() - 1) < endOfTimeWindow - math::NUMERICAL_ZERO_DIFFERENCE) {
+        data->setSampleAtTime(getTime() - _computedTimeWindowPart + _timeWindowSize, timeStepStorage.getSampleAtOrAfter(ts(ts.size() - 1)));
+      }
+      _computedTimeWindowPart = _timeWindowSize;
+    }
+  }
```

Comment:
A non-ideal place for this logic as it is in the function "did I reach the end?", where one would not expect such an alteration of the state. Better to put this in a separate function and call at the right place.

@BenjaminRodenberg Could you please double-check the logic here?

Reply:
> A non-ideal place for this logic as it is in the function "did I reach the end?", where one would not expect such an alteration of the state. Better to put this in a separate function and call at the right place.

Yes. Additionally, I would also claim that this function should be a `const` function. The way it is implemented currently this is not the case anymore.
 
> @BenjaminRodenberg Could you please double-check the logic here?

Logic is quite straight-forward. Looks ok. But I think the readability could be improved a lot and then this would also shrink down to a few lines of code. I added some comments.